File: 130.surrounded-regions.java
/*
 * @lc app=leetcode id=130 lang=java
 *
 * [130] Surrounded Regions
 */

// @lc code=start
class Solution {

    int[][] moves = new int[][] {{-1,0},{0,1},{1,0},{0,-1}};

    public void solve(char[][] board) {
        if (board == null || board.length == 0 || board[0].length == 0) return;
        boolean[][] visited = new boolean[board.length][board[0].length];
        for (int i = 0; i < board.length; ++i) {
            dfs(board, visited, i, 0);
            dfs(board, visited, i, board[0].length-1);
        }
        for (int j = 1; j < board[0].length-1; ++j) {
            dfs(board, visited, 0, j);
            dfs(board, visited, board.length-1, j);
        }
        for (int i = 0; i < board.length; ++i) {
            for (int j = 0; j < board[0].length; ++j) {
                if (board[i][j] == 'O' && !visited[i][j]) {
                    board[i][j] = 'X';
                }
            }
        }
    }

    private void dfs(char[][] board, boolean[][] visited, int i, int j) {
        if (board[i][j] != 'O' || visited[i][j]) return;
        visited[i][j] = true;
        for (int[] move : moves) {
            int ni = i + move[0], nj = j + move[1];
            if (ni < 0 || ni >= board.length || nj < 0 || nj >= board[0].length) {
                continue;
            }
            dfs(board, visited, ni, nj);
        }
    }
}
// @lc code=end

File: go/130.surrounded-regions.go
/*
 * @lc app=leetcode id=130 lang=golang
 *
 * [130] Surrounded Regions
 */

// @lc code=start
func solve(board [][]byte)  {
	visited := make([][]bool, len(board))
	for i := 0; i < len(board); i++ {
		visited[i] = make([]bool, len(board[0]))
	}

	var dfs func(int, int)
	dfs = func(i, j int) {
		invalid := i < 0 || i >= len(board)
		invalid = invalid || j < 0 || j >= len(board[0])
		invalid = invalid || board[i][j] == 'X' || visited[i][j]
		if invalid {
			return
		}
		visited[i][j] = true
		dfs(i-1, j)
		dfs(i+1, j)
		dfs(i, j-1)
		dfs(i, j+1)
	}

	for i := 0; i < len(board); i++ {
		if i == 0 || i == len(board) - 1 {
			for j := 0; j < len(board[0]); j++ {
				if board[i][j] == 'O' && !visited[i][j] {
					dfs(i, j)
				}
			}
		} else {
			for _, j := range []int{0, len(board[0])-1} {
				if board[i][j] == 'O' && !visited[i][j] {
					dfs(i, j)
				}
			}
		}
	}
	for i := 1; i < len(board)-1; i++ {
		for j := 1; j < len(board[0])-1; j++ {
			if board[i][j] == 'O' && !visited[i][j] {
				board[i][j] = 'X'
			}
		}
	}
}
// @lc code=end

